bulkcopy.go

```diff
@@ -0,0 +1,576 @@
+		}
+	}
+	if b.Debug {
+		log.Printf("row[%d] %s\n", b.numRows, logcol)
+	}
+
+	return buf.Bytes(), nil
+
+}
+
+func (b *MssqlBulk) Done() (rowcount int64, err error) {
+	if b.headerSent == false {
+		//no rows had been sent
+		return 0, nil
+	}
+	var buf = b.cn.sess.buf
+	buf.WriteByte(tokenDone)
+
+	binary.Write(buf, binary.LittleEndian, uint16(doneFinal))
+	binary.Write(buf, binary.LittleEndian, uint16(0)) //     curcmd
+
+	if b.cn.sess.loginAck.TDSVersion >= verTDS72 {
+		binary.Write(buf, binary.LittleEndian, uint64(0)) //rowcount 0
+	} else {
+		binary.Write(buf, binary.LittleEndian, uint32(0)) //rowcount 0
+	}
+
+	buf.FinishPacket()
+
+	//
+	tokchan := make(chan tokenStruct, 5)
+	go processResponse(context.Background(), b.cn.sess, tokchan)
+
+	var rowCount int64
+	for token := range tokchan {
+		switch token := token.(type) {
+		case doneStruct:
+			if token.Status&doneCount != 0 {
+				rowCount = int64(token.RowCount)
+			}
+		case error:
+			return 0, token
+		}
+	}
+	return rowCount, nil
+}
+
+func (b *MssqlBulk) createColMetadata() []byte {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(tokenColMetadata)                                    // token
+	binary.Write(buf, binary.LittleEndian, uint16(len(b.bulkColumns))) // column count
+
+	for i, col := range b.bulkColumns {
+
+		if b.cn.sess.loginAck.TDSVersion >= verTDS72 {
+			binary.Write(buf, binary.LittleEndian, uint32(col.UserType)) //  usertype, always 0?
+		} else {
+			binary.Write(buf, binary.LittleEndian, uint16(col.UserType))
+		}
+		binary.Write(buf, binary.LittleEndian, uint16(col.Flags))
+
+		writeTypeInfo(buf, &b.bulkColumns[i].ti)
+
+		if col.ti.TypeId == typeNText ||
+			col.ti.TypeId == typeText ||
+			col.ti.TypeId == typeImage {
+
+			tablename_ucs2 := str2ucs2(b.tablename)
+			binary.Write(buf, binary.LittleEndian, uint16(len(tablename_ucs2)/2))
+			buf.Write(tablename_ucs2)
+		}
+		colname_ucs2 := str2ucs2(col.ColName)
+		buf.WriteByte(uint8(len(colname_ucs2) / 2))
+		buf.Write(colname_ucs2)
+	}
+
+	return buf.Bytes()
+}
+
+func (b *MssqlBulk) getMetadata() (err error) {
+
+	stmt, err := b.cn.Prepare("SET FMTONLY ON")
+	if err != nil {
+		return
+	}
+
+	_, err = stmt.Exec(nil)
+	if err != nil {
+		return
+	}
+
+	//get columns info
+	stmt, err = b.cn.Prepare(fmt.Sprintf("select * from %s SET FMTONLY OFF", b.tablename))
+	if err != nil {
+		return
+	}
+	stmt2 := stmt.(*MssqlStmt)
+	cols, err := stmt2.QueryMeta()
+	if err != nil {
+		return fmt.Errorf("get columns info failed: %v", err.Error())
+	}
+	b.metadata = cols
+
+	if b.Debug {
+		for _, col := range b.metadata {
+			log.Printf("col: %s typeId: %#x size: %d scale: %d prec: %d flags: %d collation: %#x\n", col.ColName, col.ti.TypeId, col.ti.Size, col.ti.Scale, col.ti.Prec, col.Flags, col.ti.Collation.lcidAndFlags)
+		}
+	}
+
+	return nil
+}
+
+//almost the same as MssqlStmt.Query, but return all columns info
+func (s *MssqlStmt) QueryMeta() (cols []columnStruct, err error) {
```

Comment:
The function comment needs to start with:
// QueryMeta is almost the same as MssqlStmt.Query, but returns all the columns info.

Remove all commented out code in this function.